### PR TITLE
Add structs for Upgrades and Forks structs

### PR DESCRIPTION
**Pull request by ValarDragon**
Closes: #1375 

## What is the purpose of the change

Simplifies a lot of the upgrades management logic / boilerplate. The goal is that every new upgrade should from the app.go level, just require a one line update. And all the complexity would otherwise be held to within each upgrades module.

This PR is marked as draft until Changelog update + docs update to the `app/upgrades` README.

## Brief change log

* Introduces an Upgrade and Fork struct
* Adapt existing upgrades to use it

## Testing and Verifying

This should be covered by the existing test for the v4 upgrade, which tests that upgrade handler execution happens.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? Needs to be Done
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -83,6 +84,9 @@ var (
 	EmptyWasmOpts []wasm.Option
 
 	_ App = (*OsmosisApp)(nil)
+
+	Upgrades []upgrades.Upgrade = []upgrades.Upgrade{v4.Upgrade, v5.Upgrade, v7.Upgrade, v8.Upgrade}
```
Comment (alexanderbez) [inline app/app.go:88]: you should be able to remove the type declarations.

File: app/forks.go [hunk 2]
```diff
@@ -2,22 +2,13 @@ package app
...
-
-	case v6.UpgradeHeight:
-		v6.RunForkLogic(ctx)
-
-	default:
-		// do nothing
-		return
+	for _, fork := range Forks {
+		if ctx.BlockHeight() == fork.UpgradeHeight {
+			fork.BeginForkLogic(ctx, &app.AppKeepers)
```
Comment (alexanderbez) [inline app/forks.go:11]: maybe consider adding a `return` just to be safe?

**Review by alexanderbez — APPROVED** [review]
Nice work!!!

File: app/upgrades/README.md [hunk 3]
```diff
@@ -10,4 +14,31 @@ This folder contains logic for every osmosis upgrade. (Both state migrations, an
...
+## Upgrade types 
+
+There are two upgrade types exposed, `Upgrade` and `Fork`. 
+An `Upgrade` defines an upgrade that is to be acted upon by state migrations from the SDK `x/upgrade` module.
+A `Fork` defines a hard fork that changes some logic at a block height. 
+If the goal is to have a new binary be compatible with the old binary prior to the upgrade height,
+as is the case for all osmosis `Fork`s, then all logic changes must be height-gated or in the `BeginForkLogic` code.
+
+```go
+type Upgrade struct {
```
Comment (p0mvn) [inline app/upgrades/README.md:26]: These code snippets are likely to go out of sync with the actual code overtime

**Review by p0mvn — APPROVED** [review]

File: app/upgrades/README.md (hunk 3 above)
Comment (ValarDragon) [reply app/upgrades/README.md:26]: Hopefully they never update O_O

**Pull request merged**